cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/draft/DraftRequestDeserializer.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.cdap.datapipeline.draft;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.reflect.TypeToken;
+import io.cdap.cdap.datapipeline.service.StudioUtil;
+import io.cdap.cdap.etl.proto.v2.DataStreamsConfig;
+import io.cdap.cdap.etl.proto.v2.ETLBatchConfig;
+import io.cdap.cdap.etl.proto.v2.ETLConfig;
+
+import java.lang.reflect.Type;
+
+/**
+ * GSON deserializer to correctly deserialize the config field in {@link DraftRequest}
+ */
+public class DraftRequestDeserializer implements JsonDeserializer<DraftRequest<ETLConfig>> {
+
+  @Override
+  public DraftRequest<ETLConfig> deserialize(JsonElement jsonElement, Type type,
+                                             JsonDeserializationContext context) throws JsonParseException {
+
+    // Deserialize this normally to get all the other field values
+    Gson gson = new GsonBuilder().create();
```

Comment:
Shouldn't new a Gson inside a deserializer. Use `context.deserialize` instead.

Reply:
I initially had it like that but I ran into an infinite loop since this class is registered as a deserializer in the `context`. So calling `context.deserialize` for the `DraftRequest` class would just call this method again over and over. So I changed it to create a new Gson object which isnt ideal...

The alternative is to manually extract each field and construct the `DraftRequest` object but that means this deserializer will have to be updated any time a field is added/removed/renamed in the DraftRequest class. Is this the better option?

Reply:
Well, one way to do it is to deserialize it twice without using a custom deserializer, which you sort of doing it already. Something look like this:

```java
DraftRequest<ETLConfig> deserialize(String requestBody) {
  DraftRequest<ETLConfig> request = gson.fromJson(requestBody, new TypeToken<DraftRequest<ETLConfig>>() { }.getType());
  if (StudioUtil.isBatchPipeline(request.getArtifact())) {
    return gson.fromJson(requestStr, new TypeToken<DraftRequest<ETLBatchConfig>>() { }.getType());
  } else if (StudioUtil.isStreamingPipeline(request.getArtifact())) {
    return gson.fromJson(requestStr, new TypeToken<DraftRequest<DataStreamsConfig>>() { }.getType());
  }
}
```

Reply:
Sounds good, I've removed the custom deserializer and added a utility method in `DraftHandler` to handle the deserialization.